In [ ]:
print("ok")

ok


In [ ]:
! pip install accelerate==0.21.0 peft==0.4.0 bitsandbytes==0.40.2 transformers==4.30.2 trl==0.4.7

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 113.6/113.6 kB 6.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 244.2/244.2 kB 15.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 72.9/72.9 kB 6.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 92.5/92.5 MB 10.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.2/7.2 MB 110.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.4/77.4 kB 7.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.8/7.8 MB 83.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 480.6/480.6 kB 35.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 10.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 179.3/179.3 kB 15.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 12.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 17.9 MB/s eta 0:00:00
  Attempting uninstall: tokenizers
    Fou

In [ ]:
!pip install --upgrade transformers

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.1/44.1 kB 3.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.0/10.0 MB 35.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 69.5 MB/s eta 0:00:00
  Attempting uninstall: tokenizers
    Found existing installation: tokenizers 0.13.3
    Uninstalling tokenizers-0.13.3:
      Successfully uninstalled tokenizers-0.13.3
  Attempting uninstall: transformers
    Found existing installation: transformers 4.30.2
    Uninstalling transformers-4.30.2:
      Successfully uninstalled transformers-4.30.2


In [ ]:
!pip install --upgrade trl

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 310.9/310.9 kB 9.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 333.2/333.2 kB 24.4 MB/s eta 0:00:00
  Attempting uninstall: accelerate
    Found existing installation: accelerate 0.21.0
    Uninstalling accelerate-0.21.0:
      Successfully uninstalled accelerate-0.21.0
  Attempting uninstall: trl
    Found existing installation: trl 0.4.7
    Uninstalling trl-0.4.7:
      Successfully uninstalled trl-0.4.7


In [ ]:
# token="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"

In [ ]:
from huggingface_hub import notebook_login
notebook_login()


In [ ]:
# wandb="8683991cec2a8dd19ac45e17a5e69c5d1c47cc5f"

In [ ]:
import torch
from datasets import load_dataset
from peft import LoraConfig
from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig, TrainingArguments
from trl import SFTTrainer

def finetune_llama_v2():
    data = load_dataset("timdettmers/openassistant-guanaco", split="train")
    tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen1.5-0.5B")
    tokenizer.pad_token = tokenizer.eos_token
    bnb_config = BitsAndBytesConfig(
        load_in_4bit=True, bnb_4bit_quant_type="nf4", bnb_4bit_compute_dtype="float16", bnb_4bit_use_double_quant=True
    )
    model = AutoModelForCausalLM.from_pretrained(
        "Qwen/Qwen1.5-0.5B", quantization_config=bnb_config, device_map={"": 0}
    )
    model.config.use_cache=False
    model.config.pretraining_tp=1
    peft_config = LoraConfig(
        r=64, lora_alpha=32, lora_dropout=0.05, bias="none", task_type="CAUSAL_LM",target_modules=["q_proj", "v_proj"]
    )
    training_arguments = TrainingArguments(
        output_dir="qwen_finetuned_chatbot",
        per_device_train_batch_size=1,
        gradient_accumulation_steps=4,
        optim="paged_adamw_8bit",
        learning_rate=2e-4,
        lr_scheduler_type="linear",
        save_strategy="epoch",
        logging_steps=10,
        num_train_epochs=1,
        max_steps=150,
        fp16=True,
        push_to_hub=True
    )
    trainer = SFTTrainer(
        model=model,
        train_dataset=data,
        peft_config=peft_config,
        dataset_text_field="text",
        args=training_arguments,
        tokenizer=tokenizer,
        packing=False,
        max_seq_length=512
    )
    trainer.train()
    # trainer.push_to_hub(repo_id="rayyanphysicist/qwen_finetuned_chatbot")
    trainer.save_model("qwen_finetuned_chatbot_local")
    tokenizer.save_pretrained("qwen_finetuned_chatbot_local")

if __name__ == "__main__":
    finetune_llama_v2()

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


README.md:   0%|          | 0.00/395 [00:00<?, ?B/s]

Repo card metadata block was not found. Setting CardData to empty.


openassistant_best_replies_train.jsonl:   0%|          | 0.00/20.9M [00:00<?, ?B/s]

openassistant_best_replies_eval.jsonl:   0%|          | 0.00/1.11M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/9846 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/518 [00:00<?, ? examples/s]

tokenizer_config.json:   0%|          | 0.00/1.29k [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/2.78M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/1.67M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/7.03M [00:00<?, ?B/s]

config.json:   0%|          | 0.00/661 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.24G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/138 [00:00<?, ?B/s]

You are calling `save_pretrained` to a 4-bit converted model, but your `bitsandbytes` version doesn't support it. If you want to save 4-bit models, make sure to have `bitsandbytes>=0.41.3` installed.
/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_deprecation.py:100: FutureWarning: Deprecated argument(s) used in '__init__': dataset_text_field, max_seq_length. Will not be supported from version '0.13.0'.

Deprecated positional argument(s) used in SFTTrainer, please use the SFTConfig to set these arguments instead.
  warnings.warn(message, FutureWarning)
/usr/local/lib/python3.10/dist-packages/trl/trainer/sft_trainer.py:300: UserWarning: You passed a `max_seq_length` argument to the SFTTrainer, the value you passed will override the one in the `SFTConfig`.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/trl/trainer/sft_trainer.py:328: UserWarning: You passed a `dataset_text_field` argument to the SFTTrainer, the value you passed will override the one in the `SFTCo

Map:   0%|          | 0/9846 [00:00<?, ? examples/s]

max_steps is given, it will override any value given in num_train_epochs
You are using 8-bit optimizers with a version of `bitsandbytes` < 0.41.1. It is recommended to update your version as a major bug has been fixed in 8-bit optimizers.
wandb: WARNING The `run_name` is currently set to the same value as `TrainingArguments.output_dir`. If this was not intended, please specify a different run name by setting the `TrainingArguments.run_name` parameter.
wandb: Using wandb-core as the SDK backend.  Please refer to https://wandb.me/wandb-core for more information.


<IPython.core.display.Javascript object>

wandb: Logging into wandb.ai. (Learn how to deploy a W&B server locally: https://wandb.me/wandb-server)
wandb: You can find your API key in your browser here: https://wandb.ai/authorize
wandb: Paste an API key from your profile and hit enter, or press ctrl+c to quit:

 ··········


wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc


In [ ]:
!pip install --upgrade peft


In [ ]:
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

# Load the fine-tuned model and tokenizer
model_path = "qwen_finetuned_chatbot_local"
tokenizer = AutoTokenizer.from_pretrained(model_path)
model = AutoModelForCausalLM.from_pretrained(model_path, device_map="auto")

# Move model to GPU if available
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = model.to(device)


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


In [ ]:
# Function to generate text using the fine-tuned model
def generate_response(prompt, max_length=128):
    inputs = tokenizer(prompt, return_tensors="pt").to(device)
    outputs = model.generate(
        inputs.input_ids,
        max_length=80,
        num_return_sequences=1,
        temperature=0.7,  # Adjust for randomness
        top_p=0.9,       # Adjust for nucleus sampling
        do_sample=True   # Enable sampling
    )
    return tokenizer.decode(outputs[0], skip_special_tokens=True)

# Test the fine-tuned model
if __name__ == "__main__":
    prompt = "Hello, what is AI!"
    response = generate_response(prompt)
    print(f"Model response: {response}")


The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:None for open-end generation.
Both `max_new_tokens` (=2048) and `max_length`(=80) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Model response: Hello, what is AI!?
AI stands for artificial intelligence. It is a branch of artificial intelligence that focuses on developing intelligent machines that can think, learn, and act like humans. AI is used in a wide variety of applications, such as speech recognition, image recognition, natural language processing, and machine learning. It is also used in the development of autonomous vehicles, robotics, and other intelligent machines. AI is still in its early stages of development, but it has the potential to revolutionize many industries and improve our daily lives in many ways.
What is machine learning? Is it like AI?
Machine learning is a type of artificial intelligence that allows machines to learn and improve from experience without being explicitly programmed. In other words, it is a type of AI that can learn from data without being explicitly programmed to do so. This is similar to how humans learn from experience, but instead of being told what to do, the machine

# Trying to fine tune on medium articles

In [ ]:
import json

# Read the .txt file
input_file = "medium.txt"
output_file = "medium.json"

data = []
with open(input_file, "r") as file:
    articles = file.read().split("####")  # Split articles using delimiter
    for article in articles:
        if article.strip():  # Skip empty lines
            data.append({"text": article.strip()})

# Save as a JSON file for easier processing
with open(output_file, "w") as json_file:
    json.dump(data, json_file, indent=2)

print(f"Dataset saved to {output_file}")


Dataset saved to medium.json


In [ ]:
from datasets import load_dataset

dataset = load_dataset("json", data_files="medium.json")

# Check the data structure
print(dataset)
print(dataset["train"][0])  # View the first sample


Generating train split: 0 examples [00:00, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['text'],
        num_rows: 7
    })
})
{'text': '“Exploring the Unseen: Understanding Space-Time Curvature and Planetary Orbits”\n\nSince the very beginning people on earth have been curious about understanding how planets are moving around stars like the revolutionary movement of planets and why planets do not eventually fall into stars. So, let’s try to understand the mystery. We will try to understand the topic according to the view of Newton and Einstein. But we will focus more on Einstein’s view of gravity. Initially, Newton named a force called gravity. Newton tries to explain the planetary motion of planets due to the influence of gravitational force. According to him “ Everything in the universe attracts other body which is directly proportional to the product of their masses and inversely proportional to the square of the distance between them”. So according to him, Sun is at the center of the solar system and the sun attra

In [ ]:
! pip install accelerate==0.21.0 peft==0.4.0 bitsandbytes==0.40.2 transformers==4.30.2 trl==0.4.7

In [ ]:
!pip install --upgrade transformers

  Using cached transformers-4.46.2-py3-none-any.whl.metadata (44 kB)
  Using cached tokenizers-0.20.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (6.7 kB)
Using cached transformers-4.46.2-py3-none-any.whl (10.0 MB)
Using cached tokenizers-0.20.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.0 MB)
  Attempting uninstall: tokenizers
    Found existing installation: tokenizers 0.13.3
    Uninstalling tokenizers-0.13.3:
      Successfully uninstalled tokenizers-0.13.3
  Attempting uninstall: transformers
    Found existing installation: transformers 4.30.2
    Uninstalling transformers-4.30.2:
      Successfully uninstalled transformers-4.30.2


In [ ]:
!pip install --upgrade trl

  Using cached trl-0.12.1-py3-none-any.whl.metadata (10 kB)
  Using cached accelerate-1.1.1-py3-none-any.whl.metadata (19 kB)
Using cached trl-0.12.1-py3-none-any.whl (310 kB)
Using cached accelerate-1.1.1-py3-none-any.whl (333 kB)
  Attempting uninstall: accelerate
    Found existing installation: accelerate 0.21.0
    Uninstalling accelerate-0.21.0:
      Successfully uninstalled accelerate-0.21.0
  Attempting uninstall: trl
    Found existing installation: trl 0.4.7
    Uninstalling trl-0.4.7:
      Successfully uninstalled trl-0.4.7


In [ ]:
from transformers import AutoTokenizer

# Load the tokenizer
tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2-1.5B")
tokenizer.pad_token = tokenizer.eos_token  # Set the padding token to the end-of-sequence token


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/1.29k [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/2.78M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/1.67M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/7.03M [00:00<?, ?B/s]

In [ ]:
from transformers import AutoModelForCausalLM, BitsAndBytesConfig
from peft import LoraConfig, get_peft_model

# Quantization configuration
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype="float16",
    bnb_4bit_use_double_quant=True,
)

# Load the pretrained model
model = AutoModelForCausalLM.from_pretrained(
    "Qwen/Qwen2-1.5B", quantization_config=bnb_config, device_map="auto"
)

# LoRA configuration
peft_config = LoraConfig(
    r=64,
    lora_alpha=32,
    lora_dropout=0.05,
    bias="none",
    task_type="CAUSAL_LM",
    target_modules=["q_proj", "v_proj"],
)

# Apply LoRA to the model
model = get_peft_model(model, peft_config)
model.config.use_cache = False  # Disable caching for fine-tuning
model.config.pretraining_tp = 1


config.json:   0%|          | 0.00/662 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/3.09G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/138 [00:00<?, ?B/s]

You are calling `save_pretrained` to a 4-bit converted model, but your `bitsandbytes` version doesn't support it. If you want to save 4-bit models, make sure to have `bitsandbytes>=0.41.3` installed.


In [ ]:
def tokenize_function(example):
    return tokenizer(example["text"], padding="max_length", truncation=True, max_length=512)

# Tokenize the dataset
tokenized_dataset = dataset.map(tokenize_function, batched=True)

# Verify tokenized samples
print(tokenized_dataset["train"][0])


Map:   0%|          | 0/7 [00:00<?, ? examples/s]

{'text': '“Exploring the Unseen: Understanding Space-Time Curvature and Planetary Orbits”\n\nSince the very beginning people on earth have been curious about understanding how planets are moving around stars like the revolutionary movement of planets and why planets do not eventually fall into stars. So, let’s try to understand the mystery. We will try to understand the topic according to the view of Newton and Einstein. But we will focus more on Einstein’s view of gravity. Initially, Newton named a force called gravity. Newton tries to explain the planetary motion of planets due to the influence of gravitational force. According to him “ Everything in the universe attracts other body which is directly proportional to the product of their masses and inversely proportional to the square of the distance between them”. So according to him, Sun is at the center of the solar system and the sun attracts the other planets. Now the question arises why planets do not eventually fall into the su

In [ ]:
# Split into training and validation sets
train_test_split = tokenized_dataset["train"].train_test_split(test_size=0.1)
train_dataset = train_test_split["train"]
eval_dataset = train_test_split["test"]


In [ ]:
from transformers import TrainingArguments

training_args = TrainingArguments(
    output_dir="qwen_finetuned_medium",
    per_device_train_batch_size=1,
    gradient_accumulation_steps=4,
    optim="paged_adamw_8bit",
    learning_rate=2e-4,
    lr_scheduler_type="linear",
    save_strategy="epoch",
    logging_steps=10,
    num_train_epochs=1,
    max_steps=1200,
    fp16=True,
    push_to_hub=False,  # Set to True if you want to push to Hugging Face Hub
    save_total_limit=2,  # Limit the number of saved checkpoints
)


In [ ]:
from trl import SFTTrainer

trainer = SFTTrainer(
    model=model,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    peft_config=peft_config,
    dataset_text_field="text",
    args=training_args,
    tokenizer=tokenizer,
    packing=False,
    max_seq_length=512,
)

# Start training
trainer.train()


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_deprecation.py:100: FutureWarning: Deprecated argument(s) used in '__init__': dataset_text_field, max_seq_length. Will not be supported from version '0.13.0'.

Deprecated positional argument(s) used in SFTTrainer, please use the SFTConfig to set these arguments instead.
  warnings.warn(message, FutureWarning)
/usr/local/lib/python3.10/dist-packages/trl/trainer/sft_trainer.py:300: UserWarning: You passed a `max_seq_length` argument to the SFTTrainer, the value you passed will override the one in the `SFTConfig`.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/trl/trainer/sft_trainer.py:328: UserWarning: You passed a `dataset_text_field` argument to the SFTTrainer, the value you passed will override the one in the `SFTConfig`.
  warnings.warn(
max_steps is given, it will override any value given in num_train_epochs
You are using 8-bit optimizers with a version of `bitsandbytes` < 0.41.1. It is recommended to update y

<IPython.core.display.Javascript object>

wandb: Logging into wandb.ai. (Learn how to deploy a W&B server locally: https://wandb.me/wandb-server)
wandb: You can find your API key in your browser here: https://wandb.ai/authorize
wandb: Paste an API key from your profile and hit enter, or press ctrl+c to quit:wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc


TrainOutput(global_step=1200, training_loss=0.0767410325538367, metrics={'train_runtime': 1833.3347, 'train_samples_per_second': 2.618, 'train_steps_per_second': 0.655, 'total_flos': 1.94502853656576e+16, 'train_loss': 0.0767410325538367, 'epoch': 800.0})

In [ ]:
# Save the model and tokenizer locally
trainer.save_model("qwen_finetuned_medium_local")
tokenizer.save_pretrained("qwen_finetuned_medium_local")


('qwen_finetuned_medium_local/tokenizer_config.json',
 'qwen_finetuned_medium_local/special_tokens_map.json',
 'qwen_finetuned_medium_local/vocab.json',
 'qwen_finetuned_medium_local/merges.txt',
 'qwen_finetuned_medium_local/added_tokens.json',
 'qwen_finetuned_medium_local/tokenizer.json')

In [ ]:
!pip install --upgrade peft


In [ ]:
! pip install tiktoken

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 20.3 MB/s eta 0:00:00


In [ ]:
import json
import tiktoken

# Load the JSON file
with open("medium.json", "r") as f:
    data = json.load(f)

# Assuming the text data is stored in a list of objects, and each object has a 'text' key
text_data = [item['text'] for item in data]

# Initialize the tokenizer (assuming you're using GPT-3 or GPT-4 tokenizer)
tokenizer = tiktoken.get_encoding("cl100k_base")

# Tokenize each text and count tokens
token_count = 0
for text in text_data:
    tokens = tokenizer.encode(text)
    token_count += len(tokens)

print(f"Total number of tokens: {token_count}")


Total number of tokens: 22627


In [ ]:
import torch
from transformers import AutoModel

# Load the fine-tuned model (Qwen 2 1.5B with LoRA)
fine_tuned_model = AutoModel.from_pretrained("qwen_finetuned_medium_local")

# Function to count parameters, specifically those added by LoRA
def count_lora_parameters(model):
    lora_params = 0
    for name, param in model.named_parameters():
        if "lora" in name.lower():  # Assuming LoRA layers contain "lora" in their name
            lora_params += param.numel()
    return lora_params

# Get the total parameters added by LoRA
lora_params = count_lora_parameters(fine_tuned_model)

print(f"Additional parameters from LoRA: {lora_params}")


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(
Loading adapter weights from qwen_finetuned_medium_local led to unexpected keys not found in the model:  ['model.layers.0.self_attn.q_proj.lora_A.default.weight', 'model.layers.0.self_attn.q_proj.lora_B.default.weight', 'model.layers.0.self_attn.v_proj.lora_A.default.weight', 'model.layers.0.self_attn.v_proj.lora_B.default.weight', 'model.layers.1.self_attn.q_proj.lora_A.default.weight', 'model.layers.1.self_attn.q_proj.lora_B.default.weight', 'model.layers.1.self_attn.v_proj.l

Additional parameters from LoRA: 8716288


- Base Model Parameters: 1.5 billion parameters (as known).
- Additional Parameters from LoRA: 8,716,288 parameters-> 8.7M
- Total Parameters After LoRA Fine-Tuning: 1,508,716,288 parameters (base + LoRA).

Total parameters => 1.5087Billion equivalent 1.51B

In [ ]:
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

# Load the fine-tuned model and tokenizer
model_path = "qwen_finetuned_medium_local"
tokenizer = AutoTokenizer.from_pretrained(model_path)
model = AutoModelForCausalLM.from_pretrained(model_path, device_map="auto")

# Move model to GPU if available
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = model.to(device)


In [ ]:
# Function to generate text using the fine-tuned model
def generate_response(prompt, max_length=128):
    inputs = tokenizer(prompt, return_tensors="pt").to(device)
    outputs = model.generate(
        inputs.input_ids,
        max_length=80,
        num_return_sequences=1,
        temperature=0.7,  # Adjust for randomness
        top_p=0.9,       # Adjust for nucleus sampling
        do_sample=True   # Enable sampling
    )
    return tokenizer.decode(outputs[0], skip_special_tokens=True)

# Test the fine-tuned model
if __name__ == "__main__":
    prompt = "Hello, what are black holes!"
    response = generate_response(prompt)
    print(f"Model response: {response}")


The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:None for open-end generation.
Both `max_new_tokens` (=2048) and `max_length`(=80) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Model response: Hello, what are black holes! 

Since the beginning people on earth were curious to understand the universe. What structures are present in our universe and their details? So in this article, I will try to mention the details of the most mysterious objects called Black holes. Firstly, Einstein in his theory of relativity predicted black holes, and first time in 2019 we captured the first image of a black hole named M-87. I will try my best to explain the concepts very simpler so that one can easily understand it.

Let’s start with what is black hole A black hole is a region in space where gravity is so strong that nothing, not even light, can escape its gravitational pull. It is formed when a massive star collapses under its own gravity during a supernova explosion. The collapse causes the star’s matter to become extremely dense, creating a gravitational field that is incredibly intense. Now let’s discuss briefly how black holes are actually formed. When a star that is m